### PR TITLE
Refactor: Follow-up of controller-builder utils

### PR DESCRIPTION
**Pull request by yue9944882**
A few feedbacks from our usage of the new controller utils..

/cc @brendandburns

### CODE REVIEW

File: examples/src/main/java/io/kubernetes/client/examples/ControllerExample.java [hunk 1]
```diff
@@ -53,29 +53,30 @@ public static void main(String[] args) throws IOException {
...
+                                .getMetadata()
+                                .getName()
+                                .startsWith("docker-")) // optional, set onUpdate filter
+                    .withOnDeleteFilter(
+                        (V1Node deletedNode, Boolean stateUnknown) ->
+                            deletedNode
+                                .getMetadata()
+                                .getName()
+                                .startsWith("docker-")) // optional, set onDelete filter
+                    .build())
```
Comment (cizezsy) [inline examples/src/main/java/io/kubernetes/client/examples/ControllerExample.java]: `ControllerWatchBuilder.build()` will throw IllegalStateException if not call with `withWorkQueue()`. If we invoke this method, we also need to pass the same workqueue to `DefaultControllerBuilder.withWorkQueue()`. Any possible to just invoke once?

File: extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerBuilder.java [hunk 2]
```diff
@@ -19,9 +19,20 @@ public static DefaultControllerBuilder defaultBuilder(SharedInformerFactory fact
   /**
    * Controller manager builder is for building controller-manager .
    *
+   * @param factory the informer factory
    * @return the controller mananger builder
    */
-  public static ControllerManangerBuilder controllerManagerBuilder() {
-    return new ControllerManangerBuilder();
+  public static ControllerManangerBuilder controllerManagerBuilder(SharedInformerFactory factory) {
```
Comment (cizezsy) [inline extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerBuilder.java]: Manager?

File: examples/src/main/java/io/kubernetes/client/examples/ControllerExample.java (hunk 1 above)
Comment (yue9944882) [reply examples/src/main/java/io/kubernetes/client/examples/ControllerExample.java]: > workqueue to DefaultControllerBuilder.withWorkQueue(). Any possible to just invoke once?

yea that involves extra verbosity indeed. any suggestion for improvement?

File: extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerBuilder.java (hunk 2 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerBuilder.java]: oh nice catch

File: examples/src/main/java/io/kubernetes/client/examples/ControllerExample.java (hunk 1 above)
Comment (yue9944882) [reply examples/src/main/java/io/kubernetes/client/examples/ControllerExample.java]: updated, ptal

Comment (cizezsy) [reply examples/src/main/java/io/kubernetes/client/examples/ControllerExample.java]: looks great! thanks~

File: extended/src/main/java/io/kubernetes/client/extended/controller/DefaultControllerWatch.java [hunk 3]
```diff
@@ -56,25 +61,43 @@ public void setOnDeleteFilterPredicate(BiPredicate<ApiType, Boolean> onDeleteFil
...
-  @Override
-  public void onAdd(ApiType obj) {
-    if (onAddFilterPredicate == null || onAddFilterPredicate.test(obj)) {
-      workQueue.add(workKeyGenerator.apply(obj));
-    }
+  public Class<ApiType> getApiTypeClass() {
+    return apiTypeClass;
+  }
+
+  public void setApiTypeClass(Class<ApiType> apiTypeClass) {
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/controller/DefaultControllerWatch.java]: When would this ever change?

File: extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerWatchBuilder.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  private Function<ApiType, Request> workKeyGenerator;
+  private WorkQueue<Request> workQueue;
+  private Class<ApiType> apiTypeClass;
+
+  private Predicate<ApiType> onAddFilterPredicate;
+  private BiPredicate<ApiType, ApiType> onUpdateFilterPredicate;
+  private BiPredicate<ApiType, Boolean> onDeleteFilterPredicate;
+
+  ControllerWatchBuilder(Class<ApiType> apiTypeClass, WorkQueue<Request> workQueue) {
+    this.workQueue = null; // required, or throw an IllegalStateException in the end.
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerWatchBuilder.java]: Why is this? That doesn't seem to make sense...

File: extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerWatchBuilder.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  public DefaultControllerWatch<ApiType> build() throws IllegalStateException {
+    // TODO: configurable resync period
+    if (null == this.workQueue) {
+      throw new IllegalStateException(
+          "work-queue must not be null when building work-queue event handler..");
+    }
+    DefaultControllerWatch<ApiType> workQueueHandler =
+        new DefaultControllerWatch<>(apiTypeClass, workQueue, workKeyGenerator);
+    if (onAddFilterPredicate != null) {
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerWatchBuilder.java]: You already test for null in the handler, so I don't think this conditional (or the conditionals below) are needed.

**Comment by brendandburns** [comment]
Mostly minor comments.

File: extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerWatchBuilder.java (hunk 5 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerWatchBuilder.java]: cleaned up

File: extended/src/main/java/io/kubernetes/client/extended/controller/DefaultControllerWatch.java (hunk 3 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/controller/DefaultControllerWatch.java]: remove the setter

File: extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerWatchBuilder.java (hunk 4 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/controller/builder/ControllerWatchBuilder.java]: removed

**Comment by yue9944882** [comment]
review addressed, @brendandburns ptal

**Pull request merged**